### PR TITLE
Deprecate _remote function

### LINKED ISSUE #7040: Deprecate `._remote()` and replace all instances with `.options()` for remote functions and actors.
Given a remote function like

```python
@ray.remote
def f():
    pass
```

We switched from syntax like

```python
f._remote(num_cpus=2)
```

to 

```python
f.options(num_cpus=2).remote()
```

Regarding deprecating `._remote`, their are several steps

1. Start logging a warning that tells people to use `.options()` instead.
2. In a subsequent release, change the warning to an exception.
3. In a subsequent release, remove the method entirely.

### PR DESCRIPTION
**Pull request by allenanswerzq**
## Why are these changes needed?

Deprecate _remote function.

## Related issue number

Closes #7040
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

**Comment by landcold7** [comment]
I think also need to change all tests that are currently use _remote function.

**Comment by landcold7** [comment]
If there is no options to set, should we still call f.options().remote()?

File: python/ray/util/__init__.py [hunk 1]
```diff
@@ -2,14 +2,13 @@
...
-    "disable_log_once_globally",
-    "enable_periodic_logging",
-    "get_actor",
-    "iter",
-    "log_once",
-    "register_actor",
+    "ActorPool", "disable_log_once_globally", "enable_periodic_logging",
+    "get_actor", "iter", "log_once", "register_actor", "renamed_function",
+    "renamed_agent", "renamed_class", "deprecation_message",
+    "deprecation_warning", "deprecated_function"
```
Comment (landcold7) [inline python/ray/util/__init__.py]: Why the ./script/format.sh gives this kind of output?

File: ci/long_running_tests/workloads/many_drivers.py [hunk 2]
```diff
@@ -57,8 +57,8 @@ def method(self):
 for _ in range(5):
     for i in range(num_nodes):
         assert (ray.get(
-            f._remote(args=[], kwargs={{}}, resources={{str(i): 1}})) == 1)
-        actor = Actor._remote(args=[], kwargs={{}}, resources={{str(i): 1}})
+            f.options(resources={{str(i): 1}})) == 1).remote(args=[], kwargs={{}})
```
Comment (edoakes) [inline ci/long_running_tests/workloads/many_drivers.py]: Don't need to use `args=[], kwargs={}` anymore if using regular remote call. Can just pass them like a normal function call. Could you update all of the callsites?

File: python/ray/actor.py [hunk 3]
```diff
@@ -101,9 +102,17 @@ def __call__(self, *args, **kwargs):
...
-        return self._remote(args, kwargs)
+        return self._remote(args, kwargs, internal_called=True)
+
+    def _remote(self,
+                args=None,
+                kwargs=None,
+                num_return_vals=None,
+                internal_called=False):
+        if not internal_called:
+            deprecation_message("Do not call this method directly, "
```
Comment (edoakes) [inline python/ray/actor.py]: Should also point them to `.options()` to dynamically set options.

File: python/ray/remote_function.py [hunk 4]
```diff
@@ -148,8 +151,13 @@ def _remote(self,
                 memory=None,
                 object_store_memory=None,
                 resources=None,
-                max_retries=None):
+                max_retries=None,
+                internal_called=False):
```
Comment (edoakes) [inline python/ray/remote_function.py]: Why do we need this `internal_called` flag? Can we just remove all instances of calling this instead?

File: python/ray/util/deprecation.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+logger = logging.getLogger(__name__)
+
+# A constant to use for any configuration that should be deprecated
+# (to check, whether this config has actually been assigned a proper value or
+# not).
+DEPRECATED_VALUE = -1
+
+
+def deprecation_message(msg, error=None):
```
Comment (edoakes) [inline python/ray/util/deprecation.py]: Adding these seems a little bit heavy-weight to me for something like deprecating APIs (which shouldn't happen often). Can we just raise `DeprecationWarning` instead?

**Review by edoakes — COMMENTED** [review]
Thanks for doing this! Left a few comments to clean things up some more.

File: ci/long_running_tests/workloads/many_drivers.py (hunk 2 above)
Comment (landcold7) [reply ci/long_running_tests/workloads/many_drivers.py]: Ok, I will resolve all the problems soon.

File: python/ray/util/deprecation.py (hunk 5 above)
Comment (landcold7) [reply python/ray/util/deprecation.py]: @edoakes Updated, please take a look, thanks.

File: ci/long_running_tests/workloads/many_tasks.py [hunk 6]
```diff
@@ -47,8 +47,9 @@ def f(*xs):
 while True:
     for _ in range(50):
         new_constrained_ids = [
-            f._remote(args=[*ids], resources={str(i % num_nodes): 1})
-            for i in range(25)
+            f.options(resources={
+                str(i % num_nodes): 1
+            }).remote(args=[*ids]) for i in range(25)
```
Comment (edoakes) [inline ci/long_running_tests/workloads/many_tasks.py]:
```suggestion
            }).remote(*ids) for i in range(25)
```

File: ci/performance_tests/test_performance.py [hunk 7]
```diff
@@ -138,7 +138,7 @@ def warm_up_cluster(num_nodes, object_store_memory):
         for i in range(num_nodes):
             for _ in range(num_objects):
                 object_ids += [
-                    create_array._remote(args=[size], resources={str(i): 1})
+                    create_array.options(args=[size], resources={str(i): 1})
```
Comment (edoakes) [inline ci/performance_tests/test_performance.py]:
```suggestion
                    create_array.options(resources={str(i): 1}).remote(size)
```
don't we need to call `.remote()` here?

File: python/ray/remote_function.py (hunk 4 above)
Comment (edoakes) [reply python/ray/remote_function.py]: I think we can just get rid of `_remote()` entirely instead. Looks like `.remote()` just calls `._remote()` directly - we can just move the implementation to `.remote()` and leave a deprecation error in `._remote()`

File: python/ray/tests/test_basic.py [hunk 8]
```diff
@@ -1236,13 +1237,17 @@ def __init__(self):
...
+            "Custom": 1
+        }).remote(args=[0])
 
-    a2 = Actor2._remote()
-    ray.get(a2.method._remote())
+    a2 = Actor2.remote()
+    ray.get(a2.method.remote())
 
-    id1, id2, id3, id4 = a.method._remote(
+    id1, id2, id3, id4 = a.method.remote(
```
Comment (edoakes) [inline python/ray/tests/test_basic.py]: need to change this to use `.options().remote()`

**Review by edoakes — COMMENTED** [review]
I think the docstrings for `.options()` for both actor classes and remote functions need to be updated to remove the reference to `._remote()` as well.

File: python/ray/remote_function.py [hunk 9]
```diff
@@ -92,7 +93,7 @@ def __init__(self, language, function, function_descriptor, num_cpus,
         # Override task.remote's signature and docstring
         @wraps(function)
         def _remote_proxy(*args, **kwargs):
-            return self._remote(args=args, kwargs=kwargs)
+            return self._remote(args=args, kwargs=kwargs, internal_called=True)
 
         self.remote = _remote_proxy
```
Comment (landcold7) [inline python/ray/remote_function.py]: @edoakes If we remove _remote completely, then this wrapper here that used to keep function's signature also needs to be removed. for detail see https://github.com/ray-project/ray/pull/4985

File: python/ray/tests/test_basic.py (hunk 8 above)
Comment (landcold7) [reply python/ray/tests/test_basic.py]: Here `a` is Actor and the ActorMethod class which `a.method` is doesn't even have the .options function.

**Pull request closed without merging**